### ISSUE TITLE
Possible RSI bug

### ISSUE THREAD

**Issue by jorkey** [ta4j/ta4j#1116]
When averageLoss.isZero() and averageGain.isZero() RSI returns zero.
RSI must return 50 when averageLoss equals averageGain.
For loss 0.001 and gain 0.001 it returns 50, but for zeros values it returns zero.
I think it's illogical.

**Comment by TheCookieLab** [comment]
In what scenario would both the average gain _and_ average loss both be zero? Periods with price losses are counted as zero for the average gain term, and periods with price gains are counted as zero for the average loss term.  For both terms to be zero it would mean there was a price gain and loss within the same period which is contradictory.

I view the existing fail safe code of:

```
if (averageLoss.isZero()) {
            return averageGain.isZero() ? zero() : hundred();
}
```

as returning a "something is wrong" signal rather than a real RSI value, I know I would if I saw an RSI of 0 or 100. Returning 50 would mask such underlying problems resulting in false signals.

**Comment by TheCookieLab** [comment]
An alternative is to return NaN, making the "exception"explicit. Any other suggestions?

**Comment by TheCookieLab** [comment]
While there may be little to no difference on a number line, in this case it's quite literally the difference between valid and invalid. Silently swallowing signs you have something deeply wrong and instead presenting a business-as-usual signal is a poor approach, especially when there's money on the line.

**Comment by jorkey** [comment]
Why do you consider the situation when both averageGain and averageLoss are zero invalid? 
This happen when the price doesn't move in any direction.
The GainIndicator and LossIndicator both return zero when price is stable.
MMAIndicator also will return zero for  zero values.

**Comment by TheCookieLab** [comment]
After mulling this over I've come around to returning 50 in this scenario. @jorkey would you mind creating a PR?

**Comment by jorkey** [comment]
Unfortunately, I can't edit XLS on my Mac. This is necessary to correct the test data.

**Comment by DamienMiras** [comment]
Bellow I displayed the highchart RSI (wich is right) and the ta4j RSI (yellow curve) on top of each other. There is a divergence and we can see that sometimes it drop to zero  where it should be at 50. Seems that problem is linked. Ineed to investigate ( this is the dayly data from  B T C kraken between oct 23 and jul 24) The divergence is probably due to the scaling due to the plujnge to 0
![image](https://github.com/user-attachments/assets/ecc0fed2-6e5a-413c-8e19-0525034ef7fd)

**Comment by DamienMiras** [comment]
> Unfortunately, I can't edit XLS on my Mac. This is necessary to correct the test data.

please could you provide a unit test or a sample programm with data that expose the bug. 
Is your data paginated ? the RSI can be zero for the first perdiod.

**Comment by DamienMiras** [comment]
The RSI express the the ratio between the momentum gain  and loss  normalized to 100
when buy and loss are equals the ratio MUST BE 50 .
if gain == 0 and loss == 0 loss = gain 
So to respect the intention behind the rsi model, when both gain and loss are zero the RSI should be 50

from a mathematical point of view, for sure 0 divide by zero is undefined.

But from a trading bot point of view the curve will go from 50.00001 straigh to zero and get up again. For sure it could be a buy signal despite the price is just doing nothing , it happen all of the time like in this today price with no liquidity the price stay still, and the avgGain and avgLoss will be zero, in the ta4jlib the rsi will go straigh to zero 

![image](https://github.com/user-attachments/assets/74960268-0930-447d-a511-3173ad1782f1)

But in all other implementation for exemple highchart, or tradingview implementation the rsi will not go to zero.

if you look at the formula 
![image](https://github.com/user-attachments/assets/9408a104-5149-48ea-95bb-13a4e350e797)
 * **A**) Returning 0 is the same as solving the equation like 0/0 = 0 
 * **B**) this is as wrong as returning 50 whith a solution like 0/0 = 1 
 * **C**) returning undefined seems to be mathematicaly right
but **A** produce catastrophic results since **B** could produce a consistant continuation. It feel like it respect the intention behind the RSI indicator which express the equlibrium momentum of the bears and the bull. Both avgGain and avgLoss equal zero is when nothing move on the market for all the measured bars, so the equilibrium should be 50 if we assume that this is the intention of the indicator wich is important, and that the equation is used as a tool instead of the absolute truth. If the equation is undefined, we as human could choose an arbitrary value. 50 seems great


 What I would do before providing a PR ... is find out what is the current implementation of highchart or tradingview for exemple wich have been proven to be consistant and reliable. Test with some data. But for sure I cant be satisfied to get zero as result, I think this is the worst solution, I rather get **B**  or **C** null or 50
 
 we could return "null" or throw. but what will happen to existing code ? throwing is very heavy to process and will hurt the performance.
 
 let me know what do you think about :) Thnaks for maintening this lib

**Comment by TheCookieLab** [comment]
So this is the code from TradingView's official Relative Strength Index indicator:

`rsi = down == 0 ? 100 : up == 0 ? 0 : 100 - (100 / (1 + up / down))`

Due to this I'm leaning away from the '50' value solution and more towards returning NaN. Any thoughts?

**Comment by DamienMiras** [comment]
seems good to me thanks for your attention

**Comment by TheCookieLab** [comment]
After further "enlightening" discussion with spaceMonkeySpace we will not be changing this after all. 

The current implementation is inline with prevailing conventions (i.e. TradingView) so any changes would put us out of sync with one of (if not the) biggest retail technical charting tool out there.

A separate, valid bug however is that the RSIIndicator's _getUnstableBars()_ always returns a hard-coded 0. Instead it should return the constructor's `barCount` param, and then the `calculate()` logic needs to check if the provided `index `is within the unstable period, and if so, return _something_. 

What this _something_ should be, is a rabbit hole (see https://github.com/ta4j/ta4j/issues/947 and https://github.com/ta4j/ta4j/issues/1141) in itself. The solution needs to be global in nature, hence its complexity and yet unresolved status.

Users of the _RSIIndicator_, are presumably already handling/ignoring the incorrect 0 bar unstable period as well as RSI values with index within the barCount (if not you should be). Without the aforementioned global solution for what to return during the unstable period, any fix would be one-off in nature. Rather than make clients update their code to handle an interim solution, it's better to do nothing for now.

If you feel strongly about this bug I encourage you to brainstorm and offer ideas within the above two issue links.

Finally, if you are trading instruments so illiquid that the price doesn't change for 10+ bars I hope - for your sake - that you're a professional market maker. If you're a retail trader, you're doing it wrong.

**Comment by spaceMonkeySpace** [comment]
"Finally, if you are trading instruments so illiquid that the price doesn't change for 10+ bars I hope - for your sake - that you're a professional market maker. If you're a retail trader, you're doing it wrong." 

lol the screenshot displays BTC in 5min jully 2024.

I give up this lib is poorly maintened, with poor decisions

The RSI you provide is wrapped arround EMA wich is NOT the trading view implementation you not even understand what you are doing and pretend to maintain a library full of bug and non sens like using zonedatetime as date field for each period

Just saying Macd also return 0 for unstable bar, this has been pointed to you by another user on discord.

I already switch to another library maintened by professionals who know what they are talking about wich is definetly NOT your case you are incompetent

**Comment by TheCookieLab** [comment]
I certainly could be incompetent. On other hand I'm also financially independent (self-made) largely achieved through 15+ years of successful trading and investing.  Is it possible that I have no idea what I'm doing and am simply amazingly lucky? Sure, but I don't lose any sleep over it. Farewell and good luck with your trading.

**Comment by spaceMonkeySpace** [comment]
Dishonest person tend to sleep well because they dont care, they dont care  to scam others by delibaratly let bugs happen will they are pointed out. They dont care to "front run" the market as you admit yourself you do. So they dont care they are  not lucky they just own money by scamming others.

**Comment by TheCookieLab** [comment]
Sure, if that's what helps you sleep at night.

**Comment by spaceMonkeySpace** [comment]
I am a honnest professionnal developer and trader 18 years+ in  bank and bigdata of happy trading good night good luck. hope more users understand that this lib is not a recomandation

**Comment by spaceMonkeySpace** [comment]
for sure you sleep well even in plain daylight ...have fun
